src/annotator.cpp

```diff
@@ -0,0 +1,1285 @@
+            auto v2 = v1->equivalentVariable(e);
+            if (Variable::equivalenceConnectionId(v1, v2).empty()) {
+                auto id = makeUniqueId();
+                Variable::setEquivalenceConnectionId(v1, v2, id);
+                auto v1v2 = std::make_pair(v1, v2);
+                mIdList.insert(std::make_pair(id, std::make_pair(Annotator::Type::CONNECTION, v1v2)));
+            }
+        }
+    }
+    for (size_t c = 0; c < parent->componentCount(); ++c) {
+        doSetConnectionIds(parent->component(c));
+    }
+}
+
+void Annotator::AnnotatorImpl::doSetMapVariablesIds(const ComponentPtr &parent)
+{
+    for (size_t v = 0; v < parent->variableCount(); ++v) {
+        auto v1 = parent->variable(v);
+        for (size_t e = 0; e < v1->equivalentVariableCount(); ++e) {
+            auto v2 = v1->equivalentVariable(e);
+            if (Variable::equivalenceMappingId(v1, v2).empty()) {
+                auto id = makeUniqueId();
+                Variable::setEquivalenceMappingId(v1, v2, id);
+                auto v1v2 = std::make_pair(v1, v2);
+                mIdList.insert(std::make_pair(id, std::make_pair(Annotator::Type::MAP_VARIABLES, v1v2)));
+            }
+        }
+    }
+    for (size_t c = 0; c < parent->componentCount(); ++c) {
+        doSetMapVariablesIds(parent->component(c));
+    }
+}
+
+void Annotator::AnnotatorImpl::doSetComponentRefIds(const ComponentPtr &parent)
+{
+    if (parent->encapsulationId().empty() && parent->componentCount() > 0) {
+        auto id = makeUniqueId();
+        parent->setEncapsulationId(id);
+        mIdList.insert(std::make_pair(id, std::make_pair(Annotator::Type::COMPONENT_REF, parent)));
+    }
+    for (size_t c = 0; c < parent->componentCount(); ++c) {
+        if (parent->component(c)->encapsulationId().empty()) {
+            auto id = makeUniqueId();
+            parent->component(c)->setEncapsulationId(id);
+            mIdList.insert(std::make_pair(id, std::make_pair(Annotator::Type::COMPONENT_REF, parent->component(c))));
+        }
+        doSetComponentRefIds(parent->component(c));
+    }
+}
+
+void Annotator::AnnotatorImpl::doSetEncapsulationId()
+{
+    if (mModel->encapsulationId().empty()) {
+        auto id = makeUniqueId();
+        mModel->setEncapsulationId(id);
+        mIdList.insert(std::make_pair(id, std::make_pair(Annotator::Type::ENCAPSULATION, mModel)));
+    }
+}
+
+void Annotator::AnnotatorImpl::doSetModelId()
+{
+    if (mModel->id().empty()) {
+        auto id = makeUniqueId();
+        mModel->setId(id);
+        mIdList.insert(std::make_pair(id, std::make_pair(Annotator::Type::MODEL, mModel)));
+    }
+}
+
+void Annotator::AnnotatorImpl::doSetAllAutomaticIds()
+{
+    std::string id;
+    // Traverse the model's entities and add ids where they are currently missing.
+    if (mModel->id().empty()) {
+        id = makeUniqueId();
+        mModel->setId(id);
+        mIdList.insert(std::make_pair(id, std::make_pair(Annotator::Type::MODEL, mModel)));
```

Comment:
`doSetModelId`

Reply:
(... I don't get it ...?)

Reply:
this is duplicating the code that is the `doSetModelId()` method...